File: python/remove_element.py
class Solution:
    def removeElement(self, nums: list[int], val: int) -> int:
        """time complexity: O(n)"""
        l: int = 0
        for r in range(len(nums)):
            if nums[r] == val:
                continue
            nums[l] = nums[r]
            l += 1
        return l


def test_removeElement_case_1():
    # arrange
    nums: list[int] = [3, 2, 2, 3]
    val: int = 3
    expected: int = 2

    # act
    solution = Solution()
    actual = solution.removeElement(nums, val)

    # assert
    assert expected == actual


def test_removeElement_case_2():
    # arrange
    nums: list[int] = [0, 1, 2, 2, 3, 0, 4, 2]
    val: int = 2
    expected: int = 5

    # act
    solution = Solution()
    actual = solution.removeElement(nums, val)

    # assert
    assert expected == actual

File: rust/src/code/remove_element.rs
struct Solution {}

impl Solution {
    pub fn remove_element(nums: &mut Vec<i32>, val: i32) -> i32 {
        let mut k: usize = 0;
        for i in 0..nums.len() {
            if nums[i] != val {
                nums[k] = nums[i];
                k += 1;
            }
        }

        k as i32
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_remove_element_case_1() {
        // arrange
        let mut nums: Vec<i32> = vec![3, 2, 2, 3];
        let val: i32 = 3;
        let expected: i32 = 2;

        // act
        let actual = Solution::remove_element(&mut nums, val);

        // assert
        assert_eq!(expected, actual);
    }
}
